File: app/src/main/java/com/krystianwsul/checkme/domainmodel/Schedule.java
package com.krystianwsul.checkme.domainmodel;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.util.Pair;

import com.krystianwsul.checkme.loaders.CreateTaskLoader;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.ScheduleType;
import com.krystianwsul.checkme.utils.time.ExactTimeStamp;
import com.krystianwsul.checkme.utils.time.TimeStamp;

import junit.framework.Assert;

import java.util.List;

public abstract class Schedule {
    @NonNull
    final DomainFactory mDomainFactory;

    Schedule(@NonNull DomainFactory domainFactory) {
        mDomainFactory = domainFactory;
    }

    @NonNull
    protected abstract ScheduleBridge getScheduleBridge();

    @NonNull
    ExactTimeStamp getStartExactTimeStamp() {
        return new ExactTimeStamp(getScheduleBridge().getStartTime());
    }

    public long getStartTime() {
        return getScheduleBridge().getStartTime();
    }

    @Nullable
    ExactTimeStamp getEndExactTimeStamp() {
        if (getScheduleBridge().getEndTime() == null)
            return null;
        else
            return new ExactTimeStamp(getScheduleBridge().getEndTime());
    }

    @Nullable
    public Long getEndTime() {
        return getScheduleBridge().getEndTime();
    }

    void setEndExactTimeStamp(@NonNull ExactTimeStamp endExactTimeStamp) {
        Assert.assertTrue(current(endExactTimeStamp));

        getScheduleBridge().setEndTime(endExactTimeStamp.getLong());
    }

    public boolean current(@NonNull ExactTimeStamp exactTimeStamp) {
        ExactTimeStamp startExactTimeStamp = getStartExactTimeStamp();
        ExactTimeStamp endExactTimeStamp = getEndExactTimeStamp();

        return (startExactTimeStamp.compareTo(exactTimeStamp) <= 0 && (endExactTimeStamp == null || endExactTimeStamp.compareTo(exactTimeStamp) > 0));
    }

    @NonNull
    ScheduleType getType() {
        return getScheduleBridge().getScheduleType();
    }

    @Nullable
    public abstract CustomTimeKey getCustomTimeKey();

    @NonNull
    public abstract List<Instance> getInstances(@NonNull Task task, ExactTimeStamp givenStartExactTimeStamp, @NonNull ExactTimeStamp givenExactEndTimeStamp);

    public abstract boolean isVisible(@NonNull Task task, @NonNull ExactTimeStamp now);

    @NonNull
    public abstract String getScheduleText(@NonNull Context context);

    @Nullable
    public abstract TimeStamp getNextAlarm(@NonNull ExactTimeStamp now);

    public void delete() {
        mDomainFactory.getTaskForce(getScheduleBridge().getRootTaskKey()).deleteSchedule(this);
        getScheduleBridge().delete();
    }

    @NonNull
    public abstract CreateTaskLoader.ScheduleData getScheduleData();

    @Nullable
    Pair<String, String> getRemoteCustomTimeKey() {
        return getScheduleBridge().getRemoteCustomTimeKey();
    }

    @NonNull
    public abstract ScheduleType getScheduleType();
}

File: app/src/main/java/com/krystianwsul/checkme/domainmodel/Schedule.kt
package com.krystianwsul.checkme.domainmodel

import com.krystianwsul.checkme.utils.ScheduleType
import com.krystianwsul.checkme.utils.time.*


abstract class Schedule(protected val domainFactory: DomainFactory) {

    protected abstract val scheduleBridge: ScheduleBridge

    protected val startExactTimeStamp by lazy { ExactTimeStamp(scheduleBridge.startTime) }

    val startTime by lazy { scheduleBridge.startTime }

    protected fun getEndExactTimeStamp() = scheduleBridge.endTime?.let { ExactTimeStamp(it) }

    val endTime get() = scheduleBridge.endTime

    val customTimeKey get() = scheduleBridge.customTimeKey

    val remoteCustomTimeKey get() = scheduleBridge.remoteCustomTimeKey

    abstract val scheduleType: ScheduleType

    val timePair
        get() = scheduleBridge.run {
            customTimeKey?.let { TimePair(it) } ?: TimePair(HourMinute(hour!!, minute!!))
        }

    protected val time
        get() = scheduleBridge.run {
            customTimeKey?.let { domainFactory.getCustomTime(it) }
                    ?: NormalTime(hour!!, minute!!)
        }

    fun setEndExactTimeStamp(endExactTimeStamp: ExactTimeStamp) {
        check(current(endExactTimeStamp))

        scheduleBridge.endTime = endExactTimeStamp.long
    }

    fun clearEndExactTimeStamp(now: ExactTimeStamp) {
        check(!current(now))

        scheduleBridge.endTime = null
    }

    fun current(exactTimeStamp: ExactTimeStamp): Boolean {
        val startExactTimeStamp = startExactTimeStamp
        val endExactTimeStamp = getEndExactTimeStamp()

        return startExactTimeStamp <= exactTimeStamp && (endExactTimeStamp == null || endExactTimeStamp > exactTimeStamp)
    }

    abstract fun getInstances(task: Task, givenStartExactTimeStamp: ExactTimeStamp?, givenExactEndTimeStamp: ExactTimeStamp): List<Instance>

    abstract fun isVisible(task: Task, now: ExactTimeStamp, hack24: Boolean): Boolean

    abstract fun getScheduleText(): String

    abstract fun getNextAlarm(now: ExactTimeStamp): TimeStamp?

    fun delete() {
        domainFactory.getTaskForce(scheduleBridge.rootTaskKey).deleteSchedule(this)
        scheduleBridge.delete()
    }

    val scheduleId get() = scheduleBridge.scheduleId
}
